baseServer_test.go

```diff
@@ -0,0 +1,377 @@
+	for _, agent := range agentArray {
+		_, exists := mapFound[agent.GetID()]
+		if exists {
+			t.Error("Duplicate")
+		} else {
+			mapFound[agent.GetID()] = 1
+		}
+	}
+	for id := range server.GetAgentMap() {
+		_, exists := mapFound[id]
+		if !exists {
+			t.Error("Value not in array")
+		}
+	}
+}
+
+func TestRunTurn(t *testing.T) {
+	m := make([]basePlatformSOMAS.AgentGeneratorCountPair[ITestBaseAgent], 1)
+	m[0] = basePlatformSOMAS.MakeAgentGeneratorCountPair(NewTestAgent, 1)
+	iterations := 1
+	rounds := 2
+	server := NewTestServer(m, iterations, rounds, time.Microsecond)
+	server.SetRunHandler(server)
+	server.Start()
+	if server.roundCounter != (iterations * rounds) {
+		t.Error("wrong number of rounds")
+	}
+}
+
+func TestAgentRecievesMessage(t *testing.T) {
+	m := make([]basePlatformSOMAS.AgentGeneratorCountPair[ITestBaseAgent], 1)
+	m[0] = basePlatformSOMAS.MakeAgentGeneratorCountPair(NewTestAgent, 1)
+	server := basePlatformSOMAS.CreateServer(m, 1, 1, time.Second)
+	agent1 := NewTestAgent(server)
+	testMessage := agent1.NewTestMessage()
+
+	arrayReceivers := make([]uuid.UUID, 1)
+	i := 0
+	for id, ag := range server.GetAgentMap() {
+		arrayReceivers[i] = id
+		i += 1
+		ag.SetGoal(1)
+	}
+
+	server.Initialise()
+	go server.SendMessage(testMessage, arrayReceivers)
+	server.EndAgentListeningSession()
+	for _, ag := range server.GetAgentMap() {
+		if !ag.ReceivedMessage() {
+			t.Error("Didn't Receive Message")
+		}
+	}
+}
+
+func TestWaitForMessagingToEnd(t *testing.T) {
+	//This tests the function WaitForMessagingToEnd() which serves as a synchronisation procedure to ensure the main
+	//thread waits for messaging to complete before moving the main thread on
+	numberOfMessages := 10099
+	const numAgents = 10
+	m := make([]basePlatformSOMAS.AgentGeneratorCountPair[ITestBaseAgent], 1)
+	m[0] = basePlatformSOMAS.MakeAgentGeneratorCountPair(NewTestAgent, numAgents)
+	server := basePlatformSOMAS.CreateServer(m, 1, 1, time.Second)
+	agentMap := server.GetAgentMap()
+	arrayOfIDs := make([]uuid.UUID, numAgents)
+	i := 0
+	for id, ag := range agentMap {
+		arrayOfIDs[i] = id
+		i++
+		ag.SetGoal(numberOfMessages *(numAgents))
+	}
+
+	for j := 0; j < numberOfMessages; j++ {
+		for _,ag := range server.GetAgentMap() {
+		msg := ag.NewTestMessage()
+		//server.IncrementWaitGroup()
+		go ag.SendMessage(msg, arrayOfIDs)
+	}}
+	server.EndAgentListeningSession()
+	for _, ag := range agentMap {
+
+		if !ag.ReceivedMessage() {
+			t.Errorf("agent %s recieved %d messages, expected %d\n", ag.GetID(), ag.GetCounter(), ag.GetGoal())
+		}
+	}
+
+}
+
+func TestAddAgent(t *testing.T) {
+	const numAgents = 10000
+	m := make([]basePlatformSOMAS.AgentGeneratorCountPair[ITestBaseAgent], 1)
+	m[0] = basePlatformSOMAS.MakeAgentGeneratorCountPair(NewTestAgent, numAgents)
+	server := basePlatformSOMAS.CreateServer(m, 1, 1, time.Second)
+	agent := NewTestAgent(server)
+	server.AddAgent(agent)
+	agMap := server.GetAgentMap()
+
+	if len(agMap) != numAgents+1 {
```

Comment:
this looks like an 'off-by-one' error. Why are we checking `n+1`?

Reply:
the test is initialising n agents and adding 1 via AddAgent then checking if the new map length is equal to n+1